lib/quicrq.c

```diff
@@ -1387,6 +1387,7 @@ int quicrq_prepare_to_send_on_stream(quicrq_stream_ctx_t* stream_ctx, void* cont
             (void)picoquic_provide_stream_data_buffer(context, 0, 0, 0);
             break;
         case quicrq_sending_fin:
+            // Suhas; send the fin bit on the stream (uni), clean up stream_ctx for that uni stream
             (void) picoquic_provide_stream_data_buffer(context, 0, 1, 0);
             stream_ctx->send_state = quicrq_sending_no_more;
             stream_ctx->is_local_finished = 1;
```

Comment:
As the comment says, you probably need to loop on all the uni_streams declared for that control stream context, and mark them as finished if they are not. In fact, you just need to mark them active, and when serving the callback set the FIN bit.

Reply:
"quicrq_prepare_to_send_on_stream" is only for the control stream, not for the uni streams. The "TODO" should be removed. In theory, the state can only be reached if all the uni streams have been sent, or all abandoned.

But there is an issue on how to finish the WARP session. In datagrams, the sender send to the receiver the final_group_id and final_object_id to the receiver as soon as it has obtained them from the source or from the previous relay. When the receiver has received up to that point, it closes the control stream. The sender will be notified, and it can close its own control stream. I think that will work as is, but we will have to test it.